### DISCUSSION TITLE
xbox elite 2 controller swapped buttons

### DISCUSSION THREAD

**Discussion by mercuriete** [kakra/linux#21] [General]
Hi @kakra ,
I recently give it a try your xpadneo patches for 5.15 again

This time I could make it working perfectly.

But the problem is... X and Y are swapped and RB is not working and some more problems.

I tried to update the controller firmware to 5.13.3143.0 but I had the following problem and I had to revert back to 4.8.1908.0
https://steamcommunity.com/groups/SteamClientBeta/discussions/0/5144805327263762385/

as you see, with the new firmware steam stopped to have the feature of remapping the paddles and I wanted to have that feature so I reverted the firmware.

My question is:
To whom I have to report the issue about having the buttons wrong using 4.8.1908.0 firmware?
Thanks for your work

**Comment by kakra** [comment]
These patches are outdated... I should update them.

Regarding the paddles, I think there's a misconception of them being separate game input buttons - MS never intended that but SDL tries to handle it that way (but it breaks at the protocol level). You can only reliably separately detect the paddles in profile 0 mode (that is, no LED is showing). All the other three profiles map the paddles to existing buttons - there's no separate key code for paddles. That mapping defaults to A,B,X,Y.

The problem with X, Y and RB you are seeing is probably due to Steam Input enabled. After (but not sure when) I upgraded this patchset to the latest v0.9 version, that may be fixed. Try installing the latest v0.9 xpadneo driver over this patchset as a module (it should work, I'm doing the same). If that fixes your problem, confirm here. If it doesn't, open an issue report there.

**Reply by mercuriete** [reply]
The problem is your patchset have the same module name than xpadneo and I compiled your module in-kernel. So I can't unload and load xpadneo.
I will try to recompile my kernel without your patchset and I will try what you are suggesting.
Thanks for answering!

Edit: about the paddles issue... The latest 5.x update Microsoft breaks the profile 0 and even the controller name changed to Xbox one instead of Xbox elite.
It seems that my controller updated to a regular Xbox controller and I had to revert the firmware update.

**Reply by kakra** [reply]
I've tried all supported controller models running firmware 5.x, and all buttons are working. Tho, xpadneo moves the Share button to a keyboard subdevice in latest master, and will assign it to F12 key (this will be configurable later, and will also support paddle configuration around that time). Paddles are working fine on the evdev API level with firmware 5.x. Where are you seeing problems? Please report to the xpadneo project.

With firmware 5.x, all supported controllers identify as "Xbox Wireless Controller" on the bus, many user-space software uses heuristics to detect the real controller model by looking at the PID (which xpadneo unifies into a single model to easily abstract it for SDL compatibility no matter the featureset) or the number of buttons (which is inconsistent anyways). It looks like only a few user-space software uses the bus name.

With the xpadneo v0.9 branch, all buttons should be visible on evdev as native gaming input buttons, paddles and share button alike. It works here for firmware 5.13 XBE2 controller in profile 0. You can watch `dmesg -W` to see it switching profiles if you do. If no native profile switch is available, you can use Guide+A,B,X,Y to switch between profile 0-3, tho this is not yet very useful for non-paddle models.

**Reply by kakra** [reply]
BTW: Switching between the patched-in module and the out-of-tree module is currently broken and needs a reboot. This is due to differences in handling the bus detach and attach which has been repaired in latest out-of-tree module.

**Reply by kakra** [reply]
I've updated the kernel patch, please re-download and apply again. It should fix most issues and is at current xpadneo master with additional Share-button-to-F12 patch (which I'm currently testing).

**Reply by mercuriete** [reply]
Ok,
I had problems on 5.x on windows because steam stops to detect the elite controller as elite and the paddles stop working.
So I reverted back.
the swapped button problems are because I had Steam Input enabled.
I will try the new patches...
but I would want to keep my controller at firmware at 4.x
It that is not possible... I think I will lose paddle remapping on windows.

Thank for your support :)

**Reply by kakra** [reply]
Maybe we can solve the paddle mystery when I reverse-engineered the programming protocol. You may need to map the paddles to some specific inputs using the Xbox Acc app in Windows. And Steam for Windows probably needs an update to work with the latest firmware properly because MS seems to have changed how the paddles are exposed on the protocol. I think @Medusalix from xone sees a similar problem with the paddles using the dongle protocol.

It looks like some bits shifted by one byte between the 4.x and the 5.x protocol which can be detected by looking at the packet size, at least this is true for the Bluetooth HID protocol. Also, at least in HID, the paddles aren't really a button bitmap but rather a plain HID value, and I needed to parse that manually in the report handler. So, in Windows you may have better luck switching from Bluetooth to the GIP dongle, or vice-versa.

With firmware 5.x, it's at least easy to switch the connection mode, no re-pairing needed:

- double tap the connect button to switch to GIP mode (dongle)
- hold the connect button for 2-3 seconds to switch to HID mode (Bluetooth)
- connect a USB cable to switch to GIP mode (USB)

**Reply by mercuriete** [reply]
I don't remember having pairing issues with 4.x because I share the same key on both operative systems:
https://unix.stackexchange.com/questions/255509/bluetooth-pairing-on-dual-boot-of-windows-linux-mint-ubuntu-stop-having-to-p

Maybe I am wrong because I mainly use the PS4 dual shock controller on Linux and I don't play too much with my elite controller.

I think I will update to 5.x hoping Valve will update the windows xbox elite support.
Thank you very much.
